platform/ui/src/components/content/viewportDownloadForm/ViewportDownloadForm.js

```diff
@@ -62,125 +76,231 @@ const ViewportDownloadForm = ({
+    if (updatedDimension === dimensions[dimension]) {
       return;
     }
 
+    newDimensions[dimension] = updatedDimension;
+
+    if (keepAspect && newDimensions[opositeDimension] !== '') {
+      newDimensions[opositeDimension] = Math.round(newDimensions[dimension] * aspectMultiplier[opositeDimension]);
+    }
+
     // In current code, keepAspect is always `true`
     // And we always start w/ a square width/height
-    setDimensions({
-      width: updatedDimension,
-      height: updatedDimension,
-    });
+    setDimensions(newDimensions);
 
     // Only update if value is non-empty
     if (!isEmpty) {
-      setViewportElementDimensions({
-        height: updatedDimension,
-        width: updatedDimension,
-      });
+      setViewportElementDimensions(newDimensions);
       setDownloadCanvas(state => ({
         ...state,
-        height: updatedDimension,
-        width: updatedDimension,
+        ...newDimensions,
       }));
     }
   };
 
-  const downloadImage = () => {
-    downloadBlob(
-      filename || DEFAULT_FILENAME,
-      fileType,
+  const error_messages = {
+    width: t('minWidthError'),
+    height: t('minHeightError'),
+    filename: t('emptyFilenameError'),
+  };
+
+  const renderErrorHandler = errorType => {
+    if (!error[errorType]) {
+      return null;
+    }
+
+    return <div className="input-error">{error_messages[errorType]}</div>;
+  };
+
+  const onKeepAspectToggle = () => {
+    const { width, height } = dimensions;
+    const aspectMultiplier = { ...aspectMultiplier };
+    if (!keepAspect) {
+      const base = Math.min(width, height);
+      aspectMultiplier.width = width / base;
+      aspectMultiplier.height = height / base;
+      setAspectMultiplier(aspectMultiplier);
+    }
+
+    setKeepAspect(!keepAspect);
+  }
+
+  const validSize = value => (value >= minimumSize ? value : minimumSize);
+  const loadAndUpdateViewports = useCallback(async () => {
+    const { image, width: scaledWidth, height: scaledHeight } = await loadImage(
+      activeViewport,
       viewportElement,
-      downloadCanvas.ref.current
+      dimensions.width,
+      dimensions.height,
     );
-  };
+
+    toggleAnnotations(showAnnotations, viewportElement);
+
+    const scaledDimensions = {
+      height: validSize(scaledHeight),
+      width: validSize(scaledWidth),
+    }
+
+    setViewportElementDimensions(scaledDimensions);
+    setDownloadCanvas(state => ({
+      ...state,
+      ...scaledDimensions,
+    }));
+
+    const {
+      dataUrl,
+      width: viewportElementWidth,
+      height: viewportElementHeight,
+    } = await updateViewportPreview(
+      viewportElement,
+      downloadCanvas.ref.current,
+      fileType
+    );
+
+    setViewportPreview(state => ({
+      ...state,
+      src: dataUrl,
+      width: validSize(viewportElementWidth),
+      height: validSize(viewportElementHeight),
+    }));
+    /* eslint-disable-next-line react-hooks/exhaustive-deps */
```

Comment:
Avoid at all costs to disable linting of hook rules, usually, this is the source of inconsistent components which calls effects constantly because of unrelated changes. If this is too big, maybe split?